File: 2022/09/08/relativeRanks.js
/**
 * https://leetcode.com/problems/relative-ranks/
 * @author Andrew Jarombek
 * @since 9/8/2022
 */

const findRelativeRanks = function(score) {
    const sorted = [...score].sort((a, b) => b - a);
    const map = new Map();

    for (let i = 0; i < sorted.length; i++) {
        map.set(sorted[i], i + 1);
    }

    for (let i = 0; i < score.length; i++) {
        let val = `${map.get(score[i])}`;

        if (val === "1") {
            val = "Gold Medal";
        } else if (val === "2") {
            val = "Silver Medal";
        } else if (val === "3") {
            val = "Bronze Medal";
        }

        score[i] = val;
    }

    return score;
};

let result = findRelativeRanks([5, 4, 3, 2, 1]);
let expected = ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"]
console.assert(expected.every((value, index) => value === result[index]));

File: 2023/07/09/go/relative_ranks.go
/**
 * https://leetcode.com/problems/base-7/
 * Author: Andrew Jarombek
 * Date: 7/9/2023
 */

package _go

import (
	"sort"
	"strconv"
)

func findRelativeRanks(score []int) []string {
	sorted := make([]int, len(score))
	copy(sorted, score)

	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i] > sorted[j]
	})

	rankMap := make(map[int]string)

	for i, s := range sorted {
		switch i {
		case 0:
			rankMap[s] = "Gold Medal"
		case 1:
			rankMap[s] = "Silver Medal"
		case 2:
			rankMap[s] = "Bronze Medal"
		default:
			rankMap[s] = strconv.Itoa(i + 1)
		}
	}

	ranks := make([]string, len(score))
	for i, s := range score {
		ranks[i] = rankMap[s]
	}

	return ranks
}
